File: 0636-exclusive-time-of-functions/0636-exclusive-time-of-functions.py
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        call_stack = []
        res = [0] * n
        prev_start_time = 0

        for log in logs:
            func_id, op, timestamp = log.split(":")
            func_id = int(func_id)
            timestamp = int(timestamp)
            if op == "start":
                if call_stack:
                    poll = call_stack[-1]
                    res[poll] += timestamp - prev_start_time

                call_stack.append(func_id)
                prev_start_time = timestamp
            else:
                pop = call_stack.pop()
                res[pop] += timestamp - prev_start_time + 1
                prev_start_time = timestamp + 1
        return res

File: 0636-exclusive-time-of-functions/0636-exclusive-time-of-functions.ts
function exclusiveTime(n: number, logs: string[]): number[] {
    let prev_time = 0
    const stack = []
    const res = new Array(n).fill(0)
    for(const log of logs) {
        const tokens = log.split(":")
        const id = parseInt(tokens[0])
        const type = tokens[1]
        const ts = parseInt(tokens[2])

        if(type === "start") {
            prev_time++
            if(stack.length === 0) {
                stack.push(id)
            } else {
                const diff = ts - prev_time
                const last_process = stack[stack.length-1]
                res[last_process] += diff
                stack.push(id)
            }
            prev_time = ts
        } else {
            stack.pop()
            const runtime = ts - prev_time + 1
            res[id] += runtime
            prev_time = ts
        }

    }
    return res

};